README.md
[![Go Report Card](https://goreportcard.com/badge/github.com/AlessioDP/kpmenu)](https://goreportcard.com/report/github.com/AlessioDP/kpmenu) [![Travis CI](https://travis-ci.com/AlessioDP/kpmenu.svg?branch=master)](https://travis-ci.com/AlessioDP/kpmenu)
# Kpmenu
Kpmenu is a tool written in Go used to view a KeePass database via a dmenu, or rofi, menu.

## Features
*   Supports KDBX v3.1 and v4.0 (based on [gokeepasslib](https://github.com/tobischo/gokeepasslib))
*   Pretty fast database decode thanks to Go
*   Interfaced with dmenu, rofi, wofi and any custom executable
*   Customize dmenu/rofi with additional command arguments
*   Kpmenu can be started as a daemon, so you don't need to re-insert credentials
    *   By default the first instance of kpmenu will enter in daemon mode (cache option) for 60 seconds
    *   You can start a permanent daemon with `--daemon` option (it won't ask open the database)
    *   Even if the cache times out, the daemon won't be killed
*   Automatically put selected value into the clipboard (for a custom time)
    *   xsel and wl-clipboard supported
    *   A custom executable can be defined for every action (copy/paste/clean clipboard)
    *   By default it will use xsel, you can override it via config or `--clipboardTool` option
    *   Hidden password typing
*   OTP support
    * If a field have an otp key, you can generate the number
    * New OTP and old TOTP methods are supported

## Dependencies
*   `go` (compile only)

## Supports
*   `dmenu`, `rofi` and `wofi` (you can define a custom executable)
*   `xsel` and `wl-clipboard` (you can define a custom executable)

## Usage
I created kpmenu to make an easy and fast way to access into my KeePass database. These are some commands that you can do:
```bash
# Open a database
kpmenu -d path/to/database.kdbx

# Open a database with a key
kpmenu -d path/to/database.kdbx -k path/to/database.key

# Open a database (credentials taken from config) with a password and rofi
kpmenu -p "mypassword" -m rofi
```

## Installation
### From AUR
You can directly install the package [kpmenu](https://aur.archlinux.org/packages/kpmenu/).

### Compiling from source
If you do not set `$GOPATH`, go sources will be downloaded into `$HOME/go`.
```bash
# Clone repository
git clone https://github.com/AlessioDP/kpmenu
cd kpmenu

# Build
make build

# Install
sudo make install
```

## Configuration
You can set options via `config` or cli arguments.

Kpmenu will check for `$HOME/.config/kpmenu/config`, you can copy the [default one](https://github.com/AlessioDP/kpmenu/blob/master/resources/config.default) with `cp ./resources/config.default $HOME/.config/kpmenu/config`.

## Options
Options taken with `kpmenu --help`
```text
Usage of kpmenu:
      --argsEntry string              Additional arguments for dmenu at entry selection, separated by a space
      --argsField string              Additional arguments for dmenu at field selection, separated by a space
      --argsMenu string               Additional arguments for dmenu at menu selection, separated by a space
      --argsPassword string           Additional arguments for dmenu at password selection, separated by a space
      --cacheOneTime                  Cache the database only the first time
      --cacheTimeout int              Timeout of cache in seconds (default 60)
  -c, --clipboardTime int             Timeout of clipboard in seconds (0 = no timeout) (default 15)
      --clipboardTool string          Choose which clipboard tool to use (default "xsel")
      --customClipboardCopy string    Custom executable for clipboard copy
      --customClipboardPaste string   Custom executable for clipboard paste
      --customPromptEntries string    Custom executable for prompt entries
      --customPromptFields string     Custom executable for prompt fields
      --customPromptMenu string       Custom executable for prompt menu
      --customPromptPassword string   Custom executable for prompt password
      --daemon                        Start kpmenu directly as daemon
  -d, --database string               Path to the KeePass database
      --fieldOrder string             String order of fields to show on field selection (default "Password UserName URL")
      --fillBlacklist string          String of blacklisted fields that won't be shown
      --fillOtherFields               Enable fill of remaining fields (default true)
  -k, --keyfile string                Path to the database keyfile
  -m, --menu string                   Choose which menu to use (default "dmenu")
  -n, --nocache                       Disable caching of database
      --nootp                         Disable OTP handling
  -p, --password string               Password of the database
      --passwordBackground string     Color of dmenu background and text for password selection, used to hide password typing (default "black")
      --textEntry string              Label for entry selection (default "Entry")
      --textField string              Label for field selection (default "Field")
      --textMenu string               Label for menu selection (default "Select")
      --textPassword string           Label for password selection (default "Password")
  -v, --version                       Show kpmenu version
```

## License
See the [LICENSE](https://github.com/AlessioDP/kpmenu/blob/master/LICENSE) file.


CHANGELOG.md
# Changelog
## 1.4.1 (2022/02/04)
* Fixed custom clipboard executables
* Updated go modules

## 1.4.0 (2021/08/07)
* Added OTP support
* Updated go modules

## 1.3.0 (2021/04/12)
* Added support for wofi
* Added custom executables, now you can configure the executable for every prompt and clipboard action
* `-r`/`--rofi` option is not used anymore, use `-m rofi`/`--menu=rofi` instead
* Typo in logging
* Updated libraries
* Updated to go 1.16

## 1.2.1 (2019/10/14)
* Updated libraries
* Improved makefile
* Updated to go 1.13

## 1.2.0 (2019/04/16)
*   Added daemon mode (option `--daemon`)
*   Updated viper and gokeepasslib libraries

## 1.1.1 (2019/03/27)
*   Added credentials log message to debug better
*   Fixed some code typos

## 1.1.0 (2019/03/25)
*   Added support for `wl-clipboard`
*   New option `--clipboardTool` to choose which clipboard manager to use
*   Changed option `--clipboard` to `--clipboardTimeout`
*   Fixed some typos


kpmenulib/config.go
package kpmenulib

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	flag "github.com/spf13/pflag"
	"github.com/spf13/viper"
)

// Configuration is the main structure of kpmenu config
type Configuration struct {
	General    ConfigurationGeneral
	Executable ConfigurationExecutable
	Style      ConfigurationStyle
	Database   ConfigurationDatabase
	Flags      Flags
}

// ConfigurationGeneral is the sub-structure of the configuration related to general kpmenu settings
type ConfigurationGeneral struct {
	Menu             string // Which menu to use
	ClipboardTool    string // Clipboard tool to use
	ClipboardTimeout int    // Clipboard timeout before clean it
	NoCache          bool   // Flag to do not cache master password
	CacheOneTime     bool   // Cache the password only the first time you write it
	CacheTimeout     int    // Timeout of cache
	NoOTP            bool   // Flag to do not handle OTPs
	DisableAutotype  bool   // Disable autotype
	AutotypeConfirm  bool   // User must always confirm
	AutotypeNoAuto   bool   // Always prompt user to select the entry to autotype
}

// ConfigurationExecutable is the sub-structure of the configuration related to tools executed by kpmenu
type ConfigurationExecutable struct {
	CustomPromptPassword   string // Custom executable for prompt password
	CustomPromptMenu       string // Custom executable for prompt menu
	CustomPromptEntries    string // Custom executable for prompt entries
	CustomPromptFields     string // Custom executable for prompt fields
	CustomClipboardCopy    string // Custom executable for clipboard copy
	CustomClipboardPaste   string // Custom executable for clipboard paste
	CustomClipboardClean   string // Custom executable for clipboard clean
	CustomAutotypeWindowID string // Custom executable for fetching title of active window
	CustomAutotypeTyper    string // Custom executable for typing results
}

// ConfigurationStyle is the sub-structure of the configuration related to style of dmenu
type ConfigurationStyle struct {
	PasswordBackground string
	TextPassword       string
	TextMenu           string
	TextEntry          string
	TextField          string
	FormatEntry        string
	ArgsPassword       string
	ArgsMenu           string
	ArgsEntry          string
	ArgsField          string
}

// ConfigurationDatabase is the sub-structure of the configuration related to database settings
type ConfigurationDatabase struct {
	Database        string
	KeyFile         string
	Password        string
	FieldOrder      string
	FillOtherFields bool
	FillBlacklist   string
}

// Flags is the sub-structure of the configuration used to handle flags that aren't into the config file
type Flags struct {
	Daemon   bool
	Version  bool
	Autotype bool
}

// Menu tools used for prompts
const (
	PromptDmenu  = "dmenu"
	PromptRofi   = "rofi"
	PromptWofi   = "wofi"
	PromptCustom = "custom"
)

// Clipboard tools used for clipboard manager
const (
	ClipboardToolXsel        = "xsel"
	ClipboardToolWlclipboard = "wl-clipboard"
	ClipboardToolCustom      = "custom"
)

// Autotype default helpers
const (
	AutotypeWindowIdentifier = "quasiauto -title"
	AutotypeTyper            = "quasiauto"
)

// NewConfiguration initializes a new Configuration pointer
func NewConfiguration() *Configuration {
	return &Configuration{
		General: ConfigurationGeneral{
			Menu:             PromptDmenu,
			ClipboardTool:    ClipboardToolXsel,
			ClipboardTimeout: 15,
			CacheTimeout:     60,
		},
		Style: ConfigurationStyle{
			PasswordBackground: "black",
			TextPassword:       "Password",
			TextMenu:           "Select",
			TextEntry:          "Entry",
			TextField:          "Field",
			FormatEntry:        "{Title} - {UserName}",
		},
		Database: ConfigurationDatabase{
			FieldOrder:      "Password UserName URL",
			FillOtherFields: true,
		},
		Executable: ConfigurationExecutable{
			CustomAutotypeWindowID: AutotypeWindowIdentifier,
			CustomAutotypeTyper:    AutotypeTyper,
		},
	}
}

// LoadConfig loads the configuration into Configuration
func (c *Configuration) LoadConfig() error {
	// Get file from config folder
	viper.SetConfigType("toml")
	viper.SetConfigFile(filepath.Join(os.Getenv("HOME"), ".config/kpmenu/config"))

	if err := viper.ReadInConfig(); err != nil {
		log.Printf("Failed to read config file: %s", err)
	} else {
		// Unmarshal general
		if err := viper.UnmarshalKey("general", &c.General); err != nil {
			return NewErrorParseConfiguration("failed to parse config file (general): %v", err)
		}

		// Unmarshal style
		if err := viper.UnmarshalKey("style", &c.Style); err != nil {
			return NewErrorParseConfiguration("failed to parse config file (style): %v", err)
		}

		// Unmarshal database
		if err := viper.UnmarshalKey("database", &c.Database); err != nil {
			return NewErrorParseConfiguration("failed to parse config file (database): %v", err)
		}

		// Unmarshal executable
		if err := viper.UnmarshalKey("executable", &c.Executable); err != nil {
			return NewErrorParseConfiguration("failed to parse config file (executable): %v", err)
		}
	}
	return nil
}

// InitializeFlags prepare cli flags
func (c *Configuration) InitializeFlags() {
	// Flags
	flag.BoolVar(&c.Flags.Daemon, "daemon", false, "Start kpmenu directly as daemon")
	flag.BoolVarP(&c.Flags.Version, "version", "v", false, "Show kpmenu version")
	flag.BoolVar(&c.Flags.Autotype, "autotype", c.Flags.Autotype, "Initiate autotype")

	// General
	flag.StringVarP(&c.General.Menu, "menu", "m", c.General.Menu, "Choose which menu to use")
	flag.StringVar(&c.General.ClipboardTool, "clipboardTool", c.General.ClipboardTool, "Choose which clipboard tool to use")
	flag.IntVarP(&c.General.ClipboardTimeout, "clipboardTime", "c", c.General.ClipboardTimeout, "Timeout of clipboard in seconds (0 = no timeout)")
	flag.BoolVarP(&c.General.NoCache, "nocache", "n", c.General.NoCache, "Disable caching of database")
	flag.BoolVar(&c.General.CacheOneTime, "cacheOneTime", c.General.CacheOneTime, "Cache the database only the first time")
	flag.IntVar(&c.General.CacheTimeout, "cacheTimeout", c.General.CacheTimeout, "Timeout of cache in seconds")
	flag.BoolVar(&c.General.NoOTP, "nootp", c.General.NoOTP, "Disable OTP handling")
	flag.BoolVar(&c.General.DisableAutotype, "noautotype", c.General.DisableAutotype, "Disable autotype handling")
	flag.BoolVar(&c.General.AutotypeConfirm, "autotypealwaysconfirm", c.General.AutotypeConfirm, "Always confirm autotype, even when there's only 1 selection")
	flag.BoolVar(&c.General.AutotypeNoAuto, "autotypeusersel", c.General.AutotypeNoAuto, "Prompt for autotype entry instead of trying to detect by active window title")

	// Executable
	flag.StringVar(&c.Executable.CustomPromptPassword, "customPromptPassword", c.Executable.CustomPromptPassword, "Custom executable for prompt password")
	flag.StringVar(&c.Executable.CustomPromptMenu, "customPromptMenu", c.Executable.CustomPromptMenu, "Custom executable for prompt menu")
	flag.StringVar(&c.Executable.CustomPromptEntries, "customPromptEntries", c.Executable.CustomPromptEntries, "Custom executable for prompt entries")
	flag.StringVar(&c.Executable.CustomPromptFields, "customPromptFields", c.Executable.CustomPromptFields, "Custom executable for prompt fields")
	flag.StringVar(&c.Executable.CustomClipboardCopy, "customClipboardCopy", c.Executable.CustomClipboardCopy, "Custom executable for clipboard copy")
	flag.StringVar(&c.Executable.CustomClipboardPaste, "customClipboardPaste", c.Executable.CustomClipboardPaste, "Custom executable for clipboard paste")
	flag.StringVar(&c.Executable.CustomAutotypeWindowID, "customAutotypeWindowID", c.Executable.CustomAutotypeWindowID, "Custom executable for identifying active window for autotype")
	flag.StringVar(&c.Executable.CustomAutotypeTyper, "customAutotypeTyper", c.Executable.CustomAutotypeTyper, "Custom executable for autotype typer")
	flag.StringVar(&c.Executable.CustomClipboardClean, "customClipboardClean", c.Executable.CustomClipboardClean, "Custom executable for clipboard clean")

	// Style
	flag.StringVar(&c.Style.PasswordBackground, "passwordBackground", c.Style.PasswordBackground, "Color of dmenu background and text for password selection, used to hide password typing")
	flag.StringVar(&c.Style.TextPassword, "textPassword", c.Style.TextPassword, "Label for password selection")
	flag.StringVar(&c.Style.TextMenu, "textMenu", c.Style.TextMenu, "Label for menu selection")
	flag.StringVar(&c.Style.TextEntry, "textEntry", c.Style.TextEntry, "Label for entry selection")
	flag.StringVar(&c.Style.TextField, "textField", c.Style.TextField, "Label for field selection")
	flag.StringVar(&c.Style.ArgsPassword, "argsPassword", c.Style.ArgsPassword, "Additional arguments for dmenu at password selection, separated by a space")
	flag.StringVar(&c.Style.ArgsMenu, "argsMenu", c.Style.ArgsMenu, "Additional arguments for dmenu at menu selection, separated by a space")
	flag.StringVar(&c.Style.ArgsEntry, "argsEntry", c.Style.ArgsEntry, "Additional arguments for dmenu at entry selection, separated by a space")
	flag.StringVar(&c.Style.ArgsField, "argsField", c.Style.ArgsField, "Additional arguments for dmenu at field selection, separated by a space")

	// Database
	flag.StringVarP(&c.Database.Database, "database", "d", c.Database.Database, "Path to the KeePass database")
	flag.StringVarP(&c.Database.KeyFile, "keyfile", "k", c.Database.KeyFile, "Path to the database keyfile")
	flag.StringVarP(&c.Database.Password, "password", "p", c.Database.Password, "Password of the database")
	flag.StringVar(&c.Database.FieldOrder, "fieldOrder", c.Database.FieldOrder, "String order of fields to show on field selection")
	flag.BoolVar(&c.Database.FillOtherFields, "fillOtherFields", c.Database.FillOtherFields, "Enable fill of remaining fields")
	flag.StringVar(&c.Database.FillBlacklist, "fillBlacklist", c.Database.FillBlacklist, "String of blacklisted fields that won't be shown")
}

// ParseFlags parses cli flags with given arguments
func (c *Configuration) ParseFlags(args []string) {
	flag.CommandLine.Parse(args)
}

// ErrParseConfiguration is the error return if the configuration loading fails
type ErrParseConfiguration struct {
	Message       string
	OriginalError error
}

// NewErrorParseConfiguration initializes the error
func NewErrorParseConfiguration(message string, err error) ErrParseConfiguration {
	return ErrParseConfiguration{
		Message:       message,
		OriginalError: err,
	}
}

func (err ErrParseConfiguration) Error() string {
	return fmt.Sprintf(err.Message, err.OriginalError)
}
